lifecycle hooks (#21061)
docker/docs: content/manuals/compose/how-tos/lifecycle.md

---
title: Using lifecycle hooks with Compose
linkTitle: Use lifecycle hooks
weight: 20
desription: How to use lifecycle hooks with Docker Compose
keywords: cli, compose, lifecycle, hooks reference
aliases:
- /compose/hooks/
---

## Services lifecycle hooks

As Docker Compose run a service container, the actual command being ran is managed
by the combination of `entrypoint` and `command` as documented [here](https://docs.docker.com/engine/containers/run/#default-entrypoint). Those ley you manage
container setup and termination, catching SITERM signal to process a graceful shutdown. In some circumstances anyway it is easier to consider those a separate

> command to be introduced as lifecycle hooks, to be ran just after the container

Reviewer: I am not sure what you mean here - 
```
In some circumstances anyway it is easier to consider those a separate
command to be introduced as lifecycle hooks, to be ran just after the container
is started, or just before Compose stop it.
```

Could you rephrase it?

is started, or just before Compose stop it.

> Another benefit for lifecycle hooks, is that hook command can run with additional

Reviewer: Suggestion: instead of 
`Another benefit for lifecycle hooks, is that hook command can run with additional privileges`
-> 
`Another advantage of lifecycle hooks is the ability to run hook commands with elevated privileges...`

privileges, typically run as `root` or even `privileged` while the container is
configured to have lower privileges to enhance security.

### Post-Start hooks

> 

Author: > These hooks don't wait for the container's main task (entrypoint) to finish.

entrypoint never finish, as this would mean container stopped. Here entrypoint and hook run concurently, with no guarantee on when things happen

> Post-Start lifecycle hooks are ran after container has been started. There's no

Reviewer: ```
There's no guarantee on the timing hook executes while the container entrypoint is ran.
```

to 
```
The timing of the hook execution is not guaranteed while the container's entrypoint is running.
```

guarantee on the timing hook executes while the container entrypoint is ran.

In following example, a post-start hook is used to set volume ownership to a
non-root user, as a fresh new volume is always created with root ownership.

```yaml
services:
  app:
    image: backend
    user: 1001
    volumes:
      - data:/data    
    post_start:
      - command: chown -R /data 1001:1001
        user: root

volumes:
  data: {} # a docker volume is created with root ownership
```

### Pre-Stop hooks

Pre-Stop lifecycle hooks are ran before an explicit request to stop container by a 
compose command, either `docker compose down` or interrupting `docker compose up`

> with Ctrl+C. Hook won't execute if container stops on its own or is killed.

Reviewer: ```suggestion
with Ctrl+C. This hook won't execute if container finishes its execution or if it is killed.
```
